src/backend/distributed/planner/distributed_planner.c

```diff
@@ -151,10 +168,22 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 		else
 		{
 			needsDistributedPlanning = ListContainsDistributedTableRTE(rangeTableList);
+			if (needsDistributedPlanning)
+			{
+				fastPathRouterQuery = FastPathRouterQuery(parse, &distributionKeyValue);
+			}
 		}
 	}
 
-	if (needsDistributedPlanning)
+	if (fastPathRouterQuery)
```

Comment:
If found this a bit confusing since we also do need to copy the parse tree for all other distributed queries. 

Maybe:
```c
if (needsDistributedPlanning)
{
    opts.originalQuery = copyObject(parse);
}

if (!fastPathRouterQuery)
{
    ...
}
```

Reply:
This is actually needed like this, because for non fast path queries AssignRTEIdentities needs to be called before the copy. I've improved the comments to make this reasoning more clear.